drv/i2c-devices/src/mwocp68.rs

```diff
@@ -350,6 +440,262 @@ impl Mwocp68 {
+            }
+        };
+
+        let send_checksum = || -> Result<UpdateState, Error> {
+            let mut data = [0u8; 4];
+
+            let checksum = match state {
+                Some(UpdateState::WroteLastBlock { checksum }) => checksum,
+                _ => panic!(),
+            };
+
+            data[0] = CommandCode::IMAGE_CHECKSUM as u8;
+            data[1] = 2;
+            data[2] = (checksum & 0xff) as u8;
+            data[3] = ((checksum >> 8) & 0xff) as u8;
+
+            self.device
+                .write(&data)
+                .map_err(|code| Error::BadWrite { cmd: data[0], code })?;
+
+            Ok(UpdateState::SentChecksum)
+        };
+
+        let verify_checksum = || -> Result<UpdateState, Error> {
+            use BOOT_LOADER_STATUS::ChecksumSuccessful;
+
+            let status = self.get_boot_loader_status()?;
+
+            match status.get_checksum_successful() {
+                Some(ChecksumSuccessful::Successful) => {
+                    Ok(UpdateState::VerifiedChecksum)
+                }
+                Some(ChecksumSuccessful::NotSuccessful) | None => {
+                    Err(Error::ChecksumNotSuccessful)
+                }
+            }
+        };
+
+        let reboot_psu = || -> Result<UpdateState, Error> {
+            self.boot_loader_command(BootLoaderCommand::BootPSUFirmware)?;
+            Ok(UpdateState::RebootedPSU)
+        };
+
+        let verify_success = || -> Result<UpdateState, Error> {
+            Ok(UpdateState::UpdateSuccessful)
+        };
+
+        //
+        // We want to confirm that our boot loader is in the state that
+        // we think it should be in.  On the one hand, this will fail in
+        // a non-totally-unreasonable fashion if we don't check this -- but
+        // we have an opportunity to assert our in-device state and fail
+        // cleanly if it doesn't match, and it feels like we should take it.
+        //
+        let expected = match state {
+            None
+            | Some(UpdateState::WroteBootLoaderKey)
+            | Some(UpdateState::WroteProductKey)
+            | Some(UpdateState::RebootedPSU) => Mode::NotBootLoader,
+
+            Some(UpdateState::BootedBootLoader)
+            | Some(UpdateState::StartedProgramming)
+            | Some(UpdateState::WroteBlock { .. })
+            | Some(UpdateState::WroteLastBlock { .. })
+            | Some(UpdateState::SentChecksum)
+            | Some(UpdateState::VerifiedChecksum) => Mode::BootLoader,
+
+            Some(UpdateState::UpdateSuccessful) => {
+                return Err(Error::UpdateAlreadySuccessful);
+            }
+        };
+
+        if self.get_boot_loader_mode()? != expected {
+            return Err(match expected {
+                Mode::BootLoader => Error::UpdateNotInBootLoader,
+                Mode::NotBootLoader => Error::UpdateInBootLoader,
+            });
+        }
+
+        let next = match state {
```

Comment:
This is an interesting architecture!

Putting stuff into helper functions makes this block less wordy, but it means that there's potential panics in the helper functions (e.g. if `send_checksum` is called from a state other than `WroteLastBlock`).  In practice, I expect the compiler to inline + optimize out the panic paths.

One alternative would be extracting the key data in this `match`, then passing it into the helper functions (so that they don't need to unpack from `state`), e.g.
```rust
let next = match state {
    Some(UpdateState::WroteLastBlock { checksum }) => send_checksum(checksum)?,
    // ... etc
```

(very optional)

Reply:
I had actually coded it this way originally, but felt it was too cluttered.  As you say, I would expect the compiler to eliminate the potential panics...